sql/core/src/main/scala/org/apache/spark/sql/ml/FrequentItems.scala

```diff
@@ -0,0 +1,124 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package org.apache.spark.sql.ml
+
+
+import org.apache.spark.sql.catalyst.plans.logical.LocalRelation
+import org.apache.spark.sql.types.{StructType, ArrayType, StructField}
+
+import scala.collection.mutable.{Map => MutableMap}
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.{Row, DataFrame, functions}
+
+private[sql] object FrequentItems extends Logging {
+
+  /**
+   * Merge two maps of counts. Subtracts the sum of `otherMap` from `baseMap`, and fills in
+   * any emptied slots with the most frequent of `otherMap`.
+   * @param baseMap The map containing the global counts
+   * @param otherMap The map containing the counts for that partition
+   * @param maxSize The maximum number of counts to keep in memory
+   */
+  private def mergeCounts[A](
+      baseMap: MutableMap[A, Long],
+      otherMap: MutableMap[A, Long],
+      maxSize: Int): Unit = {
+    val otherSum = otherMap.foldLeft(0L) { case (sum, (k, v)) =>
+      if (!baseMap.contains(k)) sum + v else sum
+    }
+    baseMap.retain((k, v) => v > otherSum)
+    // sort in decreasing order, so that we will add the most frequent items first
+    val sorted = otherMap.toSeq.sortBy(-_._2)
+    var i = 0
+    val otherSize = sorted.length
+    while (i < otherSize && baseMap.size < maxSize) {
+      val keyVal = sorted(i)
+      baseMap += keyVal._1 -> keyVal._2
+      i += 1
+    }
+  }
+  
+
+  /**
+   * Finding frequent items for columns, possibly with false positives. Using the algorithm 
+   * described in `http://www.cs.umd.edu/~samir/498/karp.pdf`.
+   * For Internal use only.
+   *
+   * @param df The input DataFrame
+   * @param cols the names of the columns to search frequent items in
+   * @param support The minimum frequency for an item to be considered `frequent`
+   * @return A Local DataFrame with the Array of frequent items for each column.
+   */
+  private[sql] def singlePassFreqItems(
+      df: DataFrame, 
+      cols: Array[String], 
```

Comment:
If multiple columns are provided, shall we search the combination of them instead of each individually? For example, if I call

~~~scala
freqItems(Array("gender", "title"), 0.01)
~~~

I'm expecting the frequent combinations instead of each of them. The current implementation is more flexible because users can create a struct from multiple columns, and this allows to find frequent items on multiple columns in parallel. But I'm a little worried about what users expect when they call `freqItems(Array("gender", "title"))` @rxin